README.md
![Speed](images/speed.png)

Golang implementation of the Performance Co-Pilot (PCP) instrumentation API

[![Build Status](https://github.com/performancecopilot/speed/workflows/CI/badge.svg)](https://github.com/performancecopilot/speed/actions/workflows/ci.yaml)
[![Coverage Status](https://coveralls.io/repos/github/performancecopilot/speed/badge.svg?branch=main)](https://coveralls.io/github/performancecopilot/speed?branch=main)
[![GoDoc](https://godoc.org/github.com/performancecopilot/speed?status.svg)](https://godoc.org/github.com/performancecopilot/speed)
[![Go Report Card](https://goreportcard.com/badge/github.com/performancecopilot/speed)](https://goreportcard.com/report/github.com/performancecopilot/speed)
[![Mailing List](https://img.shields.io/badge/Mailing%20List-pcp-blue.svg)](https://groups.io/g/pcp)
[![Slack Team](https://img.shields.io/badge/Slack-pcp-blue.svg)](https://h7zo83mvt1.execute-api.us-west-2.amazonaws.com/Express/)
[![IRC #pcp](https://img.shields.io/badge/IRC-pcp-blue.svg)](https://web.libera.chat/#pcp)
[![Github Release](https://img.shields.io/github/release/performancecopilot/speed.svg)](https://github.com/performancecopilot/speed/releases/latest)


<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Install](#install)
  - [Prerequisites](#prerequisites)
    - [PCP](#pcp)
    - [Go](#go)
    - [Grafana](#grafana)
  - [Getting the library](#getting-the-library)
  - [Getting the examples](#getting-the-examples)
- [Walkthrough](#walkthrough)
  - [SingletonMetric](#singletonmetric)
  - [InstanceMetric](#instancemetric)
  - [Counter](#counter)
  - [CounterVector](#countervector)
  - [Gauge](#gauge)
  - [GaugeVector](#gaugevector)
  - [Timer](#timer)
  - [Histogram](#histogram)
- [Go Kit](#go-kit)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Install

### Prerequisites

#### [PCP](https://pcp.io)

Install Performance Co-Pilot on your local machine, either using prebuilt archives or by getting and building the source code. For detailed instructions, read the [page from pcp.readthedocs.io](https://pcp.readthedocs.io/en/latest/HowTos/installation/index.html).

#### [Go](https://golang.org)

Set up a go environment on your computer. For more information about these steps, please read [how to write go code](https://golang.org/doc/code.html).

- download and install go 1.6 or above from [https://golang.org/dl](https://golang.org/dl)

- set up `$GOPATH` to the root folder where you want to keep your go code

- add `$GOPATH/bin` to your `$PATH` by adding `export PATH=$GOPATH/bin:$PATH` to your shell configuration file, such as to your `.bashrc`, if using a Bourne shell variant.

#### [Grafana](https://grafana-pcp.readthedocs.io/)

The grafana-pcp plugin provides PCP metrics in the popular Grafana visualization tool.  It includes [PCP Vector](https://grafana-pcp.readthedocs.io/en/latest/screenshots.html#pcp-vector), a live datasource for metrics exposed using Performance Co-Pilot. Metrics you create with Speed are immediately visible in Grafana using this datasource.

### Getting the library

```sh
go get github.com/performancecopilot/speed
```

### Getting the examples

All examples are executable go programs. Simply doing

```sh
go get github.com/performancecopilot/speed/examples/<example name>
```

will get the example and add an executable to `$GOPATH/bin`. If it is on your path, simply doing

```sh
<example name>
```

will run the binary, running the example

## Walkthrough

There are 3 main components defined in the library, a [__Client__](https://godoc.org/github.com/performancecopilot/speed#Client), a [__Registry__](https://godoc.org/github.com/performancecopilot/speed#Registry) and a [__Metric__](https://godoc.org/github.com/performancecopilot/speed#Metric). A client is created using an application name, and the same name is used to create a memory mapped file in `PCP_TMP_DIR`. Each client contains a registry of metrics that it holds, and will publish on being activated. It also has a `SetFlag` method allowing you to set a mmv flag while a mapping is not active, to one of three values, [`NoPrefixFlag`, `ProcessFlag` and `SentinelFlag`](https://godoc.org/github.com/performancecopilot/speed#MMVFlag). The ProcessFlag is the default and reports metrics prefixed with the application name (i.e. like `mmv.app_name.metric.name`). Setting it to `NoPrefixFlag` will report metrics without being prefixed with the application name (i.e. like `mmv.metric.name`) which can lead to namespace collisions, so be sure of what you're doing.

A client can register metrics to report through 2 interfaces, the first is the `Register` method, that takes a raw metric object. The other is using `RegisterString`, that can take a string with metrics and instances to register similar to the interface in parfait, along with type, semantics and unit, in that order. A client can be activated by calling the `Start` method, deactivated by the `Stop` method. While a client is active, no new metrics can be registered but it is possible to stop existing client for metric registration.

Each client contains an instance of the `Registry` interface, which can give different information like the number of registered metrics and instance domains. It also exports methods to register metrics and instance domains.

Finally, metrics are defined as implementations of different metric interfaces, but they all implement the `Metric` interface, the different metric types defined are

### [SingletonMetric](https://godoc.org/github.com/performancecopilot/speed#SingletonMetric)

This type defines a metric with no instance domain and only one value. It __requires__ type, semantics and unit for construction, and optionally takes a couple of description strings. A simple construction

```go
metric, err := speed.NewPCPSingletonMetric(
	42,                                                             // initial value
	"simple.counter",                                               // name
	speed.Int32Type,                                                // type
	speed.CounterSemantics,                                         // semantics
	speed.OneUnit,                                                  // unit
	"A Simple Metric",                                              // short description
	"This is a simple counter metric to demonstrate the speed API", // long description
)
```

A SingletonMetric supports a `Val` method that returns the metric value and a `Set(interface{})` method that sets the metric value.

### [InstanceMetric](https://godoc.org/github.com/performancecopilot/speed#InstanceMetric)

An `InstanceMetric` is a single metric object containing multiple values of the same type for multiple instances. It also __requires__ an instance domain along with type, semantics and unit for construction, and optionally takes a couple of description strings. A simple construction

```go
indom, err := speed.NewPCPInstanceDomain(
	"Acme Products",                                          // name
	[]string{"Anvils", "Rockets", "Giant_Rubber_Bands"},      // instances
	"Acme products",                                          // short description
	"Most popular products produced by the Acme Corporation", // long description
)

...

countmetric, err := speed.NewPCPInstanceMetric(
	speed.Instances{
		"Anvils":             0,
		"Rockets":            0,
		"Giant_Rubber_Bands": 0,
	},
	"products.count",
	indom,
	speed.Uint64Type,
	speed.CounterSemantics,
	speed.OneUnit,
	"Acme factory product throughput",
	`Monotonic increasing counter of products produced in the Acme Corporation
	factory since starting the Acme production application.  Quality guaranteed.`,
)
```

An instance metric supports a `ValInstance(string)` method that returns the value as well as a `SetInstance(interface{}, string)` that sets the value of a particular instance.

### [Counter](https://godoc.org/github.com/performancecopilot/speed#Counter)

A counter is simply a PCPSingletonMetric with `Int64Type`, `CounterSemantics` and `OneUnit`.
It can optionally take a short and a long description.

A simple example

```go
c, err := speed.NewPCPCounter(0, "a.simple.counter")
```

a counter supports `Set(int64)` to set a value, `Inc(int64)` to increment by a custom delta and `Up()` to increment by 1.

### [CounterVector](https://godoc.org/github.com/performancecopilot/speed#CounterVector)

A CounterVector is a PCPInstanceMetric , with `Int64Type`, `CounterSemantics` and `OneUnit` and an instance domain created and registered on initialization, with the name `metric_name.indom`.

A simple example

```go
c, err := speed.NewPCPCounterVector(
	map[string]uint64{
		"instance1": 0,
		"instance2": 1,
	}, "another.simple.counter"
)
```

It supports `Val(string)`, `Set(uint64, string)`, `Inc(uint64, string)` and `Up(string)` amongst other things.

### [Gauge](https://godoc.org/github.com/performancecopilot/speed#Gauge)

A Gauge is a simple SingletonMetric storing float64 values, i.e. a PCP Singleton Metric with `DoubleType`, `InstantSemantics` and `OneUnit`.

A simple example

```go
g, err := speed.NewPCPGauge(0, "a.sample.gauge")
```

supports `Val()`, `Set(float64)`, `Inc(float64)` and `Dec(float64)`

### [GaugeVector](https://godoc.org/github.com/performancecopilot/speed#GaugeVector)

A Gauge Vector is a PCP instance metric with `DoubleType`, `InstantSemantics` and `OneUnit` and an autogenerated instance domain. A simple example

```go
g, err := NewPCPGaugeVector(map[string]float64{
	"instance1": 1.2,
	"instance2": 2.4,
}, "met")
```

supports `Val(string)`, `Set(float64, string)`, `Inc(float64, string)` and `Dec(float64, string)`

### [Timer](https://godoc.org/github.com/performancecopilot/speed#Timer)

A timer stores the time elapsed for different operations. __It is not compatible with PCP's elapsed type metrics__. It takes a name and a `TimeUnit` for construction.

```go
timer, err := speed.NewPCPTimer("test", speed.NanosecondUnit)
```

calling `timer.Start()` signals the start of an operation

calling `timer.Stop()` signals end of an operation and will return the total elapsed time calculated by the metric so far.

### [Histogram](https://godoc.org/github.com/performancecopilot/speed#Histogram)

A histogram implements a PCP Instance Metric that reports the `mean`, `variance` and `standard_deviation` while using a histogram backed by [codahale's hdrhistogram implementation in golang](https://github.com/HdrHistogram/hdrhistogram-go). Other than these, it also returns a custom percentile and buckets for plotting graphs. It requires a low and a high value and the number of significant figures used at the time of construction.

```
m, err := speed.NewPCPHistogram("hist", 0, 1000, 5)
```

## [Go Kit](https://gokit.io)

Go kit provides [a wrapper package](https://godoc.org/github.com/go-kit/kit/metrics/pcp) over speed that can be used for building microservices that expose metrics using PCP.

For modified versions of the examples in go-kit that use pcp to report metrics, see [suyash/kit-pcp-examples](https://github.com/suyash/kit-pcp-examples)


CHANGELOG.md

v4.0.0 / 2021-07-22
===================

  * build: fix semantic import versioning (#63)

v3.1.0 / 2021-07-16
===================

  * build: update golang dependencies to modern versions (#60)
  * build: use github actions for CI (#61)
  * build: remove vendored code (#62)
  * tests: fixes for 32-bit platforms related to integers (#57)
  * Update README with modern PCP and grafana-pcp (over Vector)

v3.0.1 / 2017-10-30
===================

  * metrics: use a global InstanceDomain for PCPHistogram, fixes #54 (#55)
  * build: add golangci-lint and run lint on CI again (#56) 

v3.0.0 / 2017-10-30
===================

  * metrics: add support for multidimensional composite metrics (#50)
  * speed: remove logging from library (#49) (**BREAKING**)
  * add mmvdump tests (#48)

v2.0.0 / 2017-04-05
===================

  * examples: use log.Fatal instead of panic in all examples (#45)
  * examples: add example which exposes Go runtime metrics (#44)
  * Replace logrus with zap (#43)
  * Port Speed to Windows (#41)

v1.0.0 / 2016-10-28
===================


client.go
package speed

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/pkg/errors"

	"github.com/performancecopilot/speed/v4/bytewriter"
)

// byte lengths of different components in an mmv file
const (
	HeaderLength         = 40
	TocLength            = 16
	Metric1Length        = 104
	Metric2Length        = 48
	ValueLength          = 32
	Instance1Length      = 80
	Instance2Length      = 24
	InstanceDomainLength = 32
	StringLength         = 256
)

// MaxV1NameLength is the maximum length for a metric/instance name
// under MMV format 1
const MaxV1NameLength = 63

// MaxDataValueSize is the maximum byte length for a stored metric value, unless it is a string
const MaxDataValueSize = 16

// EraseFileOnStop if set to true, will also delete the memory mapped file
var EraseFileOnStop = false

// Client defines the interface for a type that can talk to an instrumentation agent
type Client interface {
	// a client must contain a registry of metrics
	Registry() Registry

	// starts monitoring
	Start() error

	// Start that will panic on failure
	MustStart()

	// stop monitoring
	Stop() error

	// Stop that will panic on failure
	MustStop()

	// adds a metric to be monitored
	Register(Metric) error

	// tries to add a metric to be written and panics on error
	MustRegister(Metric)

	// adds metric from a string
	RegisterString(string, interface{}, MetricType, MetricSemantics, MetricUnit) (Metric, error)

	// tries to add a metric from a string and panics on an error
	MustRegisterString(string, interface{}, MetricType, MetricSemantics, MetricUnit) Metric
}

///////////////////////////////////////////////////////////////////////////////

func mmvFileLocation(name string) (string, error) {
	if strings.ContainsRune(name, os.PathSeparator) {
		return "", errors.New("name cannot have path separator")
	}

	tdir, present := config["PCP_TMP_DIR"]
	var loc string
	if present {
		loc = filepath.Join(rootPath, tdir)
	} else {
		loc = os.TempDir()
	}

	return filepath.Join(loc, "mmv", name), nil
}

// PCPClusterIDBitLength is the bit length of the cluster id
// for a set of PCP metrics
const PCPClusterIDBitLength = 12

// MMVFlag represents an enumerated type to represent mmv flag values
type MMVFlag int

// values for MMVFlag
const (
	NoPrefixFlag MMVFlag = 1 << iota
	ProcessFlag
	SentinelFlag
)

//go:generate stringer -type=MMVFlag

// PCPClient implements a client that can generate instrumentation for PCP
type PCPClient struct {
	mutex sync.Mutex

	loc       string  // absolute location of the mmv file
	clusterID uint32  // cluster identifier for the writer
	flag      MMVFlag // write flag

	r *PCPRegistry // current registry

	writer bytewriter.Writer

	instanceoffsetc chan int
	indomoffsetc    chan int
	metricoffsetc   chan int
	valueoffsetc    chan int
	stringoffsetc   chan int
}

// NewPCPClient initializes a new PCPClient object
func NewPCPClient(name string) (*PCPClient, error) {
	return NewPCPClientWithRegistry(name, NewPCPRegistry())
}

// NewPCPClientWithRegistry initializes a new PCPClient object with the given registry
func NewPCPClientWithRegistry(name string, registry *PCPRegistry) (*PCPClient, error) {
	fileLocation, err := mmvFileLocation(name)
	if err != nil {
		return nil, errors.Wrap(err, "could not get a location for storing MMV file")
	}

	return &PCPClient{
		loc:       fileLocation,
		r:         registry,
		clusterID: hash(name, PCPClusterIDBitLength),
		flag:      ProcessFlag,
	}, nil
}

// Registry returns a writer's registry
func (c *PCPClient) Registry() Registry {
	return c.r
}

// SetFlag sets the MMVflag for the client
func (c *PCPClient) SetFlag(flag MMVFlag) error {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if c.r.mapped {
		return errors.New("cannot set mmv flag for an active client")
	}

	c.flag = flag
	return nil
}

func (c *PCPClient) tocCount() int {
	ans := 2

	if c.r.InstanceCount() > 0 {
		ans += 2
	}

	if c.r.StringCount() > 0 {
		ans++
	}

	return ans
}

// Length returns the byte length of data in the mmv file written by the current writer
func (c *PCPClient) Length() int {
	var (
		InstanceLength = Instance1Length
		MetricLength   = Metric1Length
	)

	if c.r.version2 {
		InstanceLength = Instance2Length
		MetricLength = Metric2Length
	}

	return HeaderLength +
		(c.tocCount() * TocLength) +
		(c.r.InstanceCount() * InstanceLength) +
		(c.r.InstanceDomainCount() * InstanceDomainLength) +
		(c.r.MetricCount() * MetricLength) +
		(c.r.ValuesCount() * ValueLength) +
		(c.r.StringCount() * StringLength)
}

// Start dumps existing registry data
func (c *PCPClient) Start() error {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	l := c.Length()

	writer, err := bytewriter.NewMemoryMappedWriter(c.loc, l)
	if err != nil {
		return errors.Wrap(err, "cannot create MemoryMappedBuffer in client")
	}

	c.writer = writer
	c.start()
	c.r.mapped = true
	return nil
}

func (c *PCPClient) start() {
	var (
		InstanceLength = Instance1Length
		MetricLength   = Metric1Length
	)

	if c.r.version2 {
		InstanceLength = Instance2Length
		MetricLength = Metric2Length
	}

	c.r.indomoffset = HeaderLength + TocLength*c.tocCount()
	c.r.instanceoffset = c.r.indomoffset + InstanceDomainLength*c.r.InstanceDomainCount()
	c.r.metricsoffset = c.r.instanceoffset + InstanceLength*c.r.InstanceCount()
	c.r.valuesoffset = c.r.metricsoffset + MetricLength*c.r.MetricCount()
	c.r.stringsoffset = c.r.valuesoffset + ValueLength*c.r.ValuesCount()

	if c.r.InstanceDomainCount() > 0 {
		c.instanceoffsetc, c.indomoffsetc = make(chan int, 1), make(chan int, 1)

		c.instanceoffsetc <- c.r.instanceoffset
		c.indomoffsetc <- c.r.indomoffset
	}

	if c.r.MetricCount() > 0 {
		c.metricoffsetc, c.valueoffsetc = make(chan int, 1), make(chan int, 1)

		c.metricoffsetc <- c.r.metricsoffset
		c.valueoffsetc <- c.r.valuesoffset
	}

	if c.r.StringCount() > 0 {
		c.stringoffsetc = make(chan int, 1)
		c.stringoffsetc <- c.r.stringsoffset
	}

	genc, g2offc := make(chan int64), make(chan int)

	go c.writeHeaderBlock(genc, g2offc)

	var wg sync.WaitGroup
	wg.Add(2)

	go func() {
		c.writeTocBlock()
		wg.Done()
	}()

	go func() {
		// instance domains **have** to be written before metrics
		// as metrics need instance offsets and multiple metrics
		// can have the same indom, so they need to be cached
		c.writeInstanceDomains()
		c.writeMetrics()
		wg.Done()
	}()

	gen, g2off := <-genc, <-g2offc
	wg.Wait()

	// must *always* be the last thing to happen
	_ = c.writer.MustWriteInt64(gen, g2off)
}

func (c *PCPClient) writeHeaderBlock(genc chan int64, g2offc chan int) {
	// tag
	c.writer.MustWriteString("MMV", 0)

	var pos int

	// version
	if c.r.version2 {
		pos = c.writer.MustWriteUint32(2, 4)
	} else {
		pos = c.writer.MustWriteUint32(1, 4)
	}

	// generation
	gen := time.Now().Unix()
	pos = c.writer.MustWriteInt64(gen, pos)

	g2off := pos
	pos = c.writer.MustWriteInt64(0, pos)

	// tocCount
	pos = c.writer.MustWriteInt32(int32(c.tocCount()), pos)

	// flag mask
	pos = c.writer.MustWriteInt32(int32(c.flag), pos)

	// process identifier
	pos = c.writer.MustWriteInt32(int32(os.Getpid()), pos)

	// cluster identifier
	_ = c.writer.MustWriteUint32(c.clusterID, pos)

	// NOTE: the order here is important, should be same as in start()
	// or deadlock
	genc <- gen
	g2offc <- g2off
}

func (c *PCPClient) writeTocBlock() {
	var wg sync.WaitGroup
	tocpos := HeaderLength

	wg.Add(c.tocCount())

	// instance domains toc
	if c.r.InstanceDomainCount() > 0 {
		go func(pos int) {
			// 1 is the identifier for instance domains
			c.writeSingleToc(pos, 1, c.r.InstanceDomainCount(), c.r.indomoffset)
			wg.Done()
		}(tocpos)
		tocpos += TocLength
	}

	// instances toc
	if c.r.InstanceCount() > 0 {
		go func(pos int) {
			// 2 is the identifier for instances
			c.writeSingleToc(pos, 2, c.r.InstanceCount(), c.r.instanceoffset)
			wg.Done()
		}(tocpos)
		tocpos += TocLength
	}

	// metrics and values toc
	metricsoffset, valuesoffset := c.r.metricsoffset, c.r.valuesoffset
	if c.r.MetricCount() == 0 {
		metricsoffset, valuesoffset = 0, 0
	}

	go func(pos int) {
		// 3 is the identifier for metrics
		c.writeSingleToc(pos, 3, c.r.MetricCount(), metricsoffset)
		wg.Done()
	}(tocpos)
	tocpos += TocLength

	go func(pos int) {
		// 4 is the identifier for values
		c.writeSingleToc(pos, 4, c.r.ValuesCount(), valuesoffset)
		wg.Done()
	}(tocpos)
	tocpos += TocLength

	// strings toc
	if c.r.StringCount() > 0 {
		go func(pos int) {
			// 5 is the identifier for strings
			c.writeSingleToc(pos, 5, c.r.StringCount(), c.r.stringsoffset)
			wg.Done()
		}(tocpos)
	}

	wg.Wait()
}

func (c *PCPClient) writeSingleToc(pos, identifier, count, offset int) {
	pos = c.writer.MustWriteInt32(int32(identifier), pos)
	pos = c.writer.MustWriteInt32(int32(count), pos)
	_ = c.writer.MustWriteUint64(uint64(offset), pos)
}

func (c *PCPClient) writeInstanceDomains() {
	var wg sync.WaitGroup
	wg.Add(c.r.InstanceDomainCount())

	for _, indom := range c.r.instanceDomains {
		go func(indom *PCPInstanceDomain) {
			c.writeInstanceDomain(indom)
			wg.Done()
		}(indom)
	}

	wg.Wait()
}

func (c *PCPClient) writeInstanceDomain(indom *PCPInstanceDomain) {
	off := <-c.indomoffsetc
	c.indomoffsetc <- off + InstanceDomainLength

	InstanceLength := Instance1Length
	if c.r.version2 {
		InstanceLength = Instance2Length
	}

	inoff := off
	ioff := <-c.instanceoffsetc
	c.instanceoffsetc <- ioff + InstanceLength*indom.InstanceCount()

	var wg sync.WaitGroup
	wg.Add(indom.InstanceCount())

	off = c.writer.MustWriteUint32(indom.id, off)
	off = c.writer.MustWriteInt32(int32(indom.InstanceCount()), off)
	off = c.writer.MustWriteInt64(int64(ioff), off)

	for _, i := range indom.instances {
		go func(i *pcpInstance, offset int) {
			c.writeInstance(i, inoff, offset)
			wg.Done()
		}(i, ioff)
		ioff += InstanceLength
	}

	so, lo := 0, 0

	if indom.shortDescription != "" {
		so = <-c.stringoffsetc
		c.stringoffsetc <- so + StringLength

		c.writer.MustWriteString(indom.shortDescription, so)
	}

	if indom.longDescription != "" {
		lo = <-c.stringoffsetc
		c.stringoffsetc <- lo + StringLength

		c.writer.MustWriteString(indom.longDescription, lo)
	}

	off = c.writer.MustWriteUint64(uint64(so), off)
	_ = c.writer.MustWriteUint64(uint64(lo), off)

	wg.Wait()
}

func (c *PCPClient) writeInstance(i *pcpInstance, indomoff int, off int) {
	i.offset = off

	off = c.writer.MustWriteInt64(int64(indomoff), off)
	off = c.writer.MustWriteInt32(0, off)
	off = c.writer.MustWriteUint32(i.id, off)

	if c.r.version2 {
		soff := <-c.stringoffsetc
		c.stringoffsetc <- soff + StringLength

		c.writer.MustWriteUint64(uint64(soff), off)
		c.writer.MustWriteString(i.name, soff)
	} else {
		c.writer.MustWriteString(i.name, off)
	}
}

func (c *PCPClient) writeMetrics() {
	var wg sync.WaitGroup

	launchSingletonMetric := func(metric *pcpSingletonMetric) {
		go func() {
			c.writeSingletonMetric(metric)
			wg.Done()
		}()
	}

	launchInstanceMetric := func(metric *pcpInstanceMetric) {
		go func() {
			c.writeInstanceMetric(metric)
			wg.Done()
		}()
	}

	wg.Add(c.r.MetricCount())
	for _, m := range c.r.metrics {
		switch metric := m.(type) {
		case *PCPSingletonMetric:
			launchSingletonMetric(metric.pcpSingletonMetric)
		case *PCPCounter:
			launchSingletonMetric(metric.pcpSingletonMetric)
		case *PCPGauge:
			launchSingletonMetric(metric.pcpSingletonMetric)
		case *PCPTimer:
			launchSingletonMetric(metric.pcpSingletonMetric)
		case *PCPInstanceMetric:
			launchInstanceMetric(metric.pcpInstanceMetric)
		case *PCPCounterVector:
			launchInstanceMetric(metric.pcpInstanceMetric)
		case *PCPGaugeVector:
			launchInstanceMetric(metric.pcpInstanceMetric)
		case *PCPHistogram:
			launchInstanceMetric(metric.pcpInstanceMetric)
		}
	}

	wg.Wait()
}

func (c *PCPClient) writeSingletonMetric(m *pcpSingletonMetric) {
	var wg sync.WaitGroup
	wg.Add(2)

	doff := <-c.metricoffsetc

	go func() {
		c.writeMetricDesc(m.pcpMetricDesc, m.Indom(), doff)
		wg.Done()
	}()

	off := <-c.valueoffsetc
	c.valueoffsetc <- off + ValueLength

	go func(offset int) {
		m.update = c.writeValue(m.t, m.val, offset)
		wg.Done()
	}(off)

	off = c.writer.MustWriteInt64(int64(doff), off+MaxDataValueSize)
	_ = c.writer.MustWriteInt64(0, off)

	wg.Wait()
}

func (c *PCPClient) writeInstanceMetric(m *pcpInstanceMetric) {
	var wg sync.WaitGroup
	wg.Add(1 + m.Indom().InstanceCount())

	doff := <-c.metricoffsetc

	go func() {
		c.writeMetricDesc(m.pcpMetricDesc, m.Indom(), doff)
		wg.Done()
	}()

	for name, i := range m.indom.instances {
		off := <-c.valueoffsetc
		c.valueoffsetc <- off + ValueLength

		go func(i *instanceValue, offset int) {
			i.update = c.writeValue(m.t, i.val, offset)
			wg.Done()
		}(m.vals[name], off)

		off = c.writer.MustWriteInt64(int64(doff), off+MaxDataValueSize)
		_ = c.writer.MustWriteInt64(int64(i.offset), off)
	}

	wg.Wait()
}

func (c *PCPClient) writeMetricDesc(desc *pcpMetricDesc, indom *PCPInstanceDomain, off int) {
	if c.r.version2 {
		c.metricoffsetc <- off + Metric2Length

		noff := <-c.stringoffsetc
		c.stringoffsetc <- noff + StringLength

		off = c.writer.MustWriteUint64(uint64(noff), off)
		c.writer.MustWriteString(desc.name, noff)
	} else {
		c.metricoffsetc <- off + Metric1Length

		c.writer.MustWriteString(desc.name, off)
		off += MaxV1NameLength + 1
	}

	off = c.writer.MustWriteUint32(desc.id, off)
	off = c.writer.MustWriteInt32(int32(desc.t), off)
	off = c.writer.MustWriteInt32(int32(desc.sem), off)
	off = c.writer.MustWriteUint32(desc.u.PMAPI(), off)

	if indom != nil {
		off = c.writer.MustWriteUint32(indom.ID(), off)
	} else {
		off = c.writer.MustWriteInt32(-1, off)
	}

	off = c.writer.MustWriteInt32(0, off)

	so, lo := 0, 0

	if desc.shortDescription != "" {
		so = <-c.stringoffsetc
		c.stringoffsetc <- so + StringLength

		c.writer.MustWriteString(desc.shortDescription, so)
	}

	if desc.longDescription != "" {
		lo = <-c.stringoffsetc
		c.stringoffsetc <- lo + StringLength

		c.writer.MustWriteString(desc.longDescription, lo)
	}

	off = c.writer.MustWriteUint64(uint64(so), off)
	_ = c.writer.MustWriteUint64(uint64(lo), off)
}

func (c *PCPClient) writeValue(t MetricType, val interface{}, offset int) updateClosure {
	if t == StringType {
		pos := c.writer.MustWriteUint64(StringLength-1, offset)

		offset = <-c.stringoffsetc
		c.stringoffsetc <- offset + StringLength

		c.writer.MustWriteUint64(uint64(offset), pos)
	}

	update := newupdateClosure(offset, c.writer)
	_ = update(val)

	return update
}

// MustStart is a start that panics
func (c *PCPClient) MustStart() {
	if err := c.Start(); err != nil {
		panic(err)
	}
}

// Stop removes existing mapping and cleans up
func (c *PCPClient) Stop() error {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if !c.r.mapped {
		return errors.New("trying to stop an already stopped mapping")
	}

	c.stop()

	c.r.mapped = false

	err := c.writer.(*bytewriter.MemoryMappedWriter).Unmap(EraseFileOnStop)
	c.writer = nil
	if err != nil {
		return errors.Wrap(err, "client: error unmapping MemoryMappedBuffer")
	}

	return nil
}

func (c *PCPClient) stop() {
	c.instanceoffsetc, c.indomoffsetc = nil, nil
	c.metricoffsetc, c.valueoffsetc = nil, nil
	c.stringoffsetc = nil
}

// MustStop is a stop that panics
func (c *PCPClient) MustStop() {
	if err := c.Stop(); err != nil {
		panic(err)
	}
}

// Register is simply a shorthand for Registry().AddMetric
func (c *PCPClient) Register(m Metric) error { return c.r.AddMetric(m) }

// MustRegister is simply a Register that can panic
func (c *PCPClient) MustRegister(m Metric) {
	if err := c.Register(m); err != nil {
		panic(err)
	}
}

// RegisterIndom is simply a shorthand for Registry().AddInstanceDomain
func (c *PCPClient) RegisterIndom(indom InstanceDomain) error {
	return c.r.AddInstanceDomain(indom)
}

// MustRegisterIndom is simply a RegisterIndom that can panic
func (c *PCPClient) MustRegisterIndom(indom InstanceDomain) {
	if err := c.RegisterIndom(indom); err != nil {
		panic(err)
	}
}

// RegisterString is simply a shorthand for Registry().AddMetricByString
func (c *PCPClient) RegisterString(str string, val interface{}, t MetricType, s MetricSemantics, u MetricUnit) (Metric, error) {
	return c.r.AddMetricByString(str, val, t, s, u)
}

// MustRegisterString is simply a RegisterString that panics
func (c *PCPClient) MustRegisterString(str string, val interface{}, t MetricType, s MetricSemantics, u MetricUnit) Metric {
	if m, err := c.RegisterString(str, val, t, s, u); err != nil {
		panic(err)
	} else {
		return m
	}
}
